openmls/src/group/core_group/staged_commit.rs

```diff
@@ -427,7 +491,37 @@ impl CoreGroup {
                 self.interim_transcript_hash = state.interim_transcript_hash;
 
                 self.tree.merge_diff(state.staged_diff);
-                Some(message_secrets)
+
+                // TODO #1194: Group storage and key storage should be
+                // correlated s.t. there is no divergence between key material
+                // and group state.
+
+                let leaf_keypair = if let Some(keypair) = &state.new_leaf_keypair_option {
+                    vec![keypair.clone()]
+                } else {
+                    vec![]
+                };
+
+                // Figure out which keys we need in the new epoch.
+                let new_owned_encryption_keys = self.tree.owned_encryption_keys();
+                // From the old and new keys, keep the ones that are still relevant in the new epoch.
+                let epoch_keypairs: Vec<EncryptionKeyPair> = old_epoch_keypairs
+                    .into_iter()
+                    .chain(state.new_keypairs.into_iter())
+                    .chain(leaf_keypair.into_iter())
+                    .filter(|keypair| new_owned_encryption_keys.contains(keypair.public_key()))
+                    .collect();
+                // We should have private keys for all owned encryption keys.
+                debug_assert_eq!(new_owned_encryption_keys.len(), epoch_keypairs.len());
```

Comment:
We could throw an error here, but I would say it's a bit of a correctness thing. If the code is correct and the key store works as intended, then we should have all the keys we need. If we were to throw an error here, it should probably be a key store error? Alternatively maybe a LibraryError? But in the recent past, we have tried to avoid those where possible, so I'm not sure. What do you think?

Reply:
Fair enough. I have introduced a LibraryError there for the time being. We can still rip it out later when we move the key store access further up and when we're more confident in the correctness of our key store interactions.